t/t-batch-transfer-size.sh

```diff
@@ -0,0 +1,97 @@
+﻿#!/usr/bin/env bash
+
+. "$(dirname "$0")/testlib.sh"
+
+begin_test "batch storage upload with small batch size"
+(
+  set -e
+
+  reponame="batch-storage-upload-small-batch"
+  setup_remote_repo "$reponame"
+  clone_repo "$reponame" batch-storage-repo-upload
+
+  contents1="storage-upload-batch-1"
+  contents2="storage-upload-batch-2"
+  contents3="storage-upload-batch-3"
+  oid1="$(calc_oid "$contents1")"
+  oid2="$(calc_oid "$contents2")"
+  oid3="$(calc_oid "$contents3")"
+  printf "%s" "$contents1" > a.dat
+  printf "%s" "$contents2" > b.dat
+  printf "%s" "$contents3" > c.dat
+
+  git lfs track "*.dat"
+  git add .gitattributes a.dat b.dat c.dat
+  git commit -m "initial commit"
+
+  git config --local lfs.transfer.batchSize 1
+
+  GIT_TRACE=1 git push origin main 2>&1 | tee push.log
+  if [ "0" -ne "${PIPESTATUS[0]}" ]; then
+    echo >&2 "fatal: expected \`git push origin main\` to succeed ..."
+    exit 1
+  fi
+
+  actual_count="$(grep -c "tq: sending batch of size 1" push.log)"
+  [ "3" = "$actual_count" ]
+
+  assert_server_object "$reponame" "$oid1"
+  assert_server_object "$reponame" "$oid2"
+  assert_server_object "$reponame" "$oid3"
+)
+end_test
+
+begin_test "batch storage download with small batch size"
+(
+  set -e
+
+  reponame="batch-storage-download-small-batch"
+  setup_remote_repo "$reponame"
+  clone_repo "$reponame" batch-storage-repo-download
+
+  contents1="storage-download-batch-1"
+  contents2="storage-download-batch-2"
+  contents3="storage-download-batch-3"
+  oid1="$(calc_oid "$contents1")"
+  oid2="$(calc_oid "$contents2")"
+  oid3="$(calc_oid "$contents3")"
+  printf "%s" "$contents1" > a.dat
+  printf "%s" "$contents2" > b.dat
+  printf "%s" "$contents3" > c.dat
+
+  git lfs track "*.dat"
+  git add .gitattributes a.dat b.dat c.dat
+  git commit -m "initial commit"
+
+  git push origin main
+  assert_server_object "$reponame" "$oid1"
+  assert_server_object "$reponame" "$oid2"
+  assert_server_object "$reponame" "$oid3"
+
+  pushd ..
+    git \
+      -c "filter.lfs.process=" \
+      -c "filter.lfs.smudge=cat" \
+      -c "filter.lfs.required=false" \
+      clone "$GITSERVER/$reponame" "$reponame-assert"
+
+    cd "$reponame-assert"
+
+    git config credential.helper lfstest
+    git config --local lfs.transfer.batchSize 1
+
+    GIT_TRACE=1 git lfs pull origin main 2>&1 | tee pull.log
+    if [ "0" -ne "${PIPESTATUS[0]}" ]; then
+      echo >&2 "fatal: expected \`git lfs pull origin main\` to succeed ..."
+      exit 1
+    fi
+
+    actual_count="$(grep -c "tq: sending batch of size 1" pull.log)"
+    [ "3" = "$actual_count" ]
+
+    assert_local_object "$oid1" "${#contents1}"
+    assert_local_object "$oid2" "${#contents2}"
+    assert_local_object "$oid3" "${#contents3}"
+  popd
+)
+end_test
```

Comment:
This is nicely done!  I happen to know two tricks we can use to simplify this a bit.

First, we could use `GIT_LFS_SKIP_SMUDGE=1` to avoid fetching Git LFS objects during the `git clone` step, so that would mean you don't have to explicitly override the various `filter.lfs.*` Git configuration options.

But, we don't even need to do that here, because we can actually set the `lfs.transfer.batchSize` option right in the global Git configuration file, and then just do a `git clone` and check the results.

We don't have to worry we might be contaminating the Git configuration of the "real" global Git configuration file (e.g., the `$HOME/.gitconfig` of the person running the tests in their local development environment) because the `begin_test()` shell function [resets](https://github.com/git-lfs/git-lfs/blob/df0afdd19eaf422651b6d131e4ae8a83f6399981/t/testlib.sh#L80-L84) the `HOME` environment variable at the start of every test, and copies in a `.gitconfig` file which has been explicitly [created](https://github.com/git-lfs/git-lfs/blob/df0afdd19eaf422651b6d131e4ae8a83f6399981/t/testhelpers.sh#L601-L618) with a common set of configurations so all our tests run in the same context and are never affected by any settings which might be unique to a particular development environment.  So we're free to run `git config --global` in any test, as the change will be scoped to just that test.

So I'd make the following suggestion:

```suggestion
  cd ..
  git config --global lfs.transfer.batchSize 1

  GIT_TRACE=1 git clone "$GITSERVER/$reponame" "${reponame}-assert" 2>&1 | tee clone.log
  if [ "0" -ne "${PIPESTATUS[0]}" ]; then
    echo >&2 "fatal: expected \`git clone\` to succeed ..."
    exit 1
  fi

  actual_count="$(grep -c "tq: sending batch of size 1" clone.log)"
  [ "3" = "$actual_count" ]

  cd "${reponame}-assert"
  assert_local_object "$oid1" "${#contents1}"
  assert_local_object "$oid2" "${#contents2}"
  assert_local_object "$oid3" "${#contents3}"
)
end_test
```